### DISCUSSION TITLE
Concurrent Connections Limit

### DISCUSSION THREAD

**Discussion by lentrodev** [Azure/azure-service-bus-emulator-installer#45] [Q&A]
Hey! I have a question regarding the current maximum concurrent connections limit for the Azure Service Bus Emulator. Currently, it is capped at 10 connections, which seems restrictive for some use cases.

Here’s the scenario we’re dealing with:

- We’re running .NET Aspire tests that currently rely on an Azure Service Bus instance.
- Our goal is to replace this dependency with the emulator to make the tests fully isolated.
- When running the tests, all our services use the same instance of the emulator, and we have listeners for at least 8 topics. This setup quickly consumes 8 out of the 10 available connections, leaving little room for flexibility.
- We have a plan to use Emulator when running tests for other service, which has more than 20 Azure Functions, which makes it unusable in that scenario

So here's the question - is there any plans on increasing this limit?

**Comment by krishankumar95** [comment]
Tagging @Saglodha for comments.

**Reply by PavelHrachou** [reply]
@krishankumar95 since tagging @Saglodha passed one and a half months. Should we try another communication channel?

**Comment by Saglodha** [comment]
Hey @PavelHrachou , thanks for follow up. 

Could you please highlight if there is hard restriction to only use same emulator instance for your test setup? Also, do all of those run concurrently or in sequential order? What's the maximum number of concurrent tests you run in your setup?

**Reply by lentrodev** [reply]
Hey, yep, unfortunately this is a hard restriction due to the fact that our services are interacting with each other, so there's no ability to run tests and local development on multiple instances

**Reply by PavelHrachou** [reply]
@Saglodha I'm second to @lentrodev. I did not respond before in text because I fully share his opinion and waited for the responce. This is hard restriction and prevents integration of the emulator and Aspire in the workflow. 
If it's not possible to change this, please let us know so we can stop pursuing this initiative.

**Comment by anthony-dewhirst-dfe** [comment]
This question comes up regularly (for example: https://github.com/Azure/azure-service-bus-emulator-installer/issues/58), so I wanted to add some context from actual local‑development use.
I understand the concern about discouraging production usage of the emulator and the commercial realities behind that. Using it in production would obviously bypass the real service, and I get why that’s not something you want to enable.
However, a hard limit of 10 concurrent connections makes the emulator almost unusable for realistic local development. When people talk about “testing” here, I think most mean running the application locally in a developer environment—not automated test suites. Even modest apps can easily exceed 10 connections just by starting up.
If the goal is specifically to prevent production use, there may be better ways to do that without severely limiting local fidelity. For example:

Introducing artificial throttling (e.g. random delays)
Enforcing a message cap that requires restarting the emulator
Other non-deterministic or operational penalties that would make production use unattractive

Right now, the limit pushes us away from the emulator entirely. We previously used RabbitMQ locally, and we’re likely going to have to continue doing so. The downside is that we’re no longer developing or testing against something that behaves like Azure Service Bus, which defeats the purpose of having an emulator at all.
Yes, we do test against the real service before going live. But asking every developer to connect to a shared Azure Service Bus instance is either costly or operationally painful. We’ve tried approaches like prefixing queue names with machine identifiers and sharing a single instance, but that quickly leads to dev and prod concerns bleeding into each other.
The community feedback here seems consistent: people want a closer approximation of the real Service Bus experience locally, without accidentally enabling production use.
Is it possible to revisit this constraint with that goal in mind?

**Reply by PavelHrachou** [reply]
For us, 10 connections are exceeded at startup. So we can't use this emulator to run integration tests, and opted out for an in-house solution.

**Comment by Greybird** [comment]
Any news ? We were very happy to have an emulator, only to be left disapointed with limitations. Worse, communication is very sparse, and we lack visibility on the plans with regards to this emulator, which does not help...
